### PR TITLE
multi ssl bug

### LINKED ISSUE #1817: bug: In the case of multiple ssl certificates, some certificates are not available
### Issue description

In the case of multiple ssl certificates, some certificates are not available.

### Environment

* apisix version (cmd: `apisix version`): 1.4

* OS: centos 7

### Minimal test code / Steps to reproduce the issue
1. post some ssl cert by api admin
2. use https to access the api

### What's the actual result? (including assertion message & call stack if applicable)
respond good

### What's the expected result?
respond error

### PR DESCRIPTION
**Pull request by nic-chen**
### Summary

fix multi ssl bug

This bug affects released version 1.4

### Issues resolved

Fix #1817

### CODE REVIEW

File: t/router/radixtree-sni.t [hunk 1]
```diff
@@ -448,15 +448,15 @@ location /t {
         local ssl_key =  t.read_file("conf/cert/test2.key")
         local data = {cert = ssl_cert, key = ssl_key, sni = "*.test2.com"}
 
-        local code, body = t.test('/apisix/admin/ssl/1',
+        local code, body = t.test('/apisix/admin/ssl/2',
```
Comment (membphis) [inline t/router/radixtree-sni.t:451]: we need to remove the SSL object which the path is `/apisix/admin/ssl/2`

File: apisix/http/router/radixtree_sni.lua [hunk 2]
```diff
@@ -159,7 +158,7 @@ function _M.match_and_set(api_ctx)
     local sni_rev = sni:reverse()
     local ok = radixtree_router:dispatch(sni_rev, nil, api_ctx)
     if not ok then
-        core.log.warn("not found any valid sni configuration")
+        core.log.warn("failed to found any SSL certificate by sni: ", sni)
         return false
```
Comment (membphis) [inline apisix/http/router/radixtree_sni.lua:163]: `failed to find any SSL certificate by SNI: `

File: apisix/http/router/radixtree_sni.lua [hunk 3]
```diff
@@ -172,13 +171,13 @@ function _M.match_and_set(api_ctx)
             end
         end
         if not matched then
-            core.log.warn("not found any valid sni configuration, matched sni: ",
+            core.log.warn("failed to found any valid sni configuration, matched sni: ",
```
Comment (membphis) [inline apisix/http/router/radixtree_sni.lua]: ditoo

File: apisix/http/router/radixtree_sni.lua (hunk 2 above)
Comment (membphis) [reply apisix/http/router/radixtree_sni.lua:162]: `failed to found` to `failed to find`

File: .travis/osx_openresty_runner.sh [hunk 4]
```diff
@@ -23,8 +23,8 @@ export_or_prefix() {
 }
 
 before_install() {
-    HOMEBREW_NO_AUTO_UPDATE=1 brew install perl cpanminus etcd luarocks openresty/brew/openresty-debug redis@3.2
-    brew upgrade go
+    HOMEBREW_NO_AUTO_UPDATE=1 brew install perl cpanminus etcd luarocks openresty/brew/openresty-debug redis@4.0
+    brew install go@1.13
```
Comment (moonming) [inline .travis/osx_openresty_runner.sh:27]: Is this modification related to this ssl bug?

Comment (nic-chen) [reply .travis/osx_openresty_runner.sh:27]: no, travis test failed, just for test

File: t/router/radixtree-sni.t [hunk 5]
```diff
@@ -951,7 +951,7 @@ connected: 1
 failed to do SSL handshake: certificate host mismatch
 --- error_log
 lua ssl server name: "aa.bb.test2.com"
-failed to found any SSL certificate by SNI: aa.bb.test2.com matched SNIs: ["moc.2tset","moc.2tset.*"]
+failed to find any SSL certificate by SNI: aa.bb.test2.com matched SNIs: ["moc.2tset","moc.2tset.*"]
```
Comment (membphis) [inline t/router/radixtree-sni.t:954]: Is there a way to print the correct SNI value?

`["moc.2tset","moc.2tset.*"]` was wrong, it was reversed, it is not easy to read for us.

File: t/router/radixtree-sni.t [hunk 6]
```diff
@@ -521,7 +521,82 @@ lua ssl server name: "www.test2.com"
 
 
 
-=== TEST 11: client request: aa.bb.test2.com
+=== TEST 11: client request: test.com again
```
Comment (moonming) [inline t/router/radixtree-sni.t:524]: I think you'd better create a new test file for this PR, instead of insert one test case into the exist test cases.

Comment (moonming) [reply t/router/radixtree-sni.t:524]: for example `multi-ssl-certs.t`

Comment (nic-chen) [reply t/router/radixtree-sni.t:524]: OK

File: apisix/http/router/radixtree_sni.lua [hunk 7]
```diff
@@ -96,6 +96,10 @@ local function create_router(ssl_items)
     end
 
     core.log.info("route items: ", core.json.delay_encode(route_items, true))
+    -- for testing
```
Comment (membphis) [inline apisix/http/router/radixtree_sni.lua:99]: We can check the existing error log, we do not need this testing code.

image

File: t/router/radixtree-sni.t (hunk 6 above)
Comment (membphis) [reply t/router/radixtree-sni.t:524]: So, can we delete this new case in the file?

File: t/router/multi-ssl-certs.t [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+--- error_log
+lua ssl server name: "www.test2.com"
+we have more than 1 ssl certs now
+--- no_error_log
+[error]
+[alert]
+
+
+
+=== TEST 6: set second ssl(sni: apisix.dev)
```
Comment (membphis) [inline t/router/multi-ssl-certs.t:259]: `set second ssl` -> `set third ssl`

File: apisix/http/router/radixtree_sni.lua (hunk 7 above)
Comment (nic-chen) [reply apisix/http/router/radixtree_sni.lua:99]: the existing log is a large string, it's hard to match, and the snis are reverse.
so i thought adding a new info log is better ?

File: t/router/multi-ssl-certs.t (hunk 8 above)
Comment (nic-chen) [reply t/router/multi-ssl-certs.t:259]: OK

File: apisix/http/router/radixtree_sni.lua (hunk 7 above)
Comment (membphis) [reply apisix/http/router/radixtree_sni.lua:99]: ok, got it. we can keep it.

**Comment by membphis** [comment]
merged, many thx @nic-chen
